pkg/services/move/move_router.go

```
@@ -0,0 +1,235 @@
+	logger := zap.NewNop()
+	flag := pflag.CommandLine
+	v := viper.New()
+	err := v.BindPFlags(flag)
+	if err != nil {
+		logger.Fatal("could not bind flags", zap.Error(err))
+	}
+	v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+	v.AutomaticEnv()
+
+	return v.GetBool(cli.FeatureFlagServiceCounseling)
+}
+
+// sendToServiceCounselor makes the move available for a Service Counselor to review
+func (router moveRouter) sendToServiceCounselor() error {
+	move := router.move
+	if move.Status != models.MoveStatusDRAFT {
+		return errors.Wrap(
+			models.ErrInvalidTransition, fmt.Sprintf(
+				"Cannot move to NeedsServiceCounseling state when the Move is not in Draft status. Its current status is %s",
+				move.Status,
+			),
+		)
+	}
+	move.Status = models.MoveStatusNeedsServiceCounseling
+	now := time.Now()
+	move.SubmittedAt = &now
+
+	return nil
+}
+
+// sendNewMoveToOfficeUser makes the move available for a TOO to review
+// The Submitted status indicates to the TOO that this is a new move.
+func (router moveRouter) sendNewMoveToOfficeUser() error {
+	move := router.move
+
+	if move.Status != models.MoveStatusDRAFT {
+		return errors.Wrap(models.ErrInvalidTransition, "Submit")
+	}
+	move.Status = models.MoveStatusSUBMITTED
+	now := time.Now()
+	move.SubmittedAt = &now
+
+	// Update PPM status too
+	for i := range move.PersonallyProcuredMoves {
+		ppm := &move.PersonallyProcuredMoves[i]
+		err := ppm.Submit(now)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, ppm := range move.PersonallyProcuredMoves {
+		if ppm.Advance != nil {
+			err := ppm.Advance.Request()
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Approve makes the Move available to the Prime. The Prime cannot create
+// Service Items unless the Move is approved.
+func (router moveRouter) Approve() error {
+	move := router.move
+
+	if router.approvable() {
+		move.Status = models.MoveStatusAPPROVED
+		return nil
+	}
+	if router.alreadyApproved() {
+		return nil
```

I think we might want to not silently do nothing if we try to approve a move that is already approved.  Is this actively happening somewhere?

---

Good question. This is existing behavior. We do the same thing when setting Approvals Requested. It looks like the only places where to try to approve a move are:
- when we make it available to the prime, 
- when we create service items
- when we update service items

And for Approvals Requested, it's when we create or update service items.